### PR TITLE
[WIP] Filter pushdown

### PR DESCRIPTION
**Pull request by robert3005**
- **Filter pushdown through layouts**
- **chunked**
- **columns**

### CODE REVIEW

File: vortex-serde/Cargo.toml [hunk 1]
```diff
@@ -34,22 +36,22 @@ vortex-scalar = { workspace = true, features = ["flatbuffers"] }
...
+rstest = { workspace = true }
 simplelog = { workspace = true }
 tokio = { workspace = true, features = ["full"] }
 vortex-alp = { path = "../encodings/alp" }
 vortex-fastlanes = { path = "../encodings/fastlanes" }
 vortex-sampling-compressor = { path = "../vortex-sampling-compressor" }
 
 [lints]
-workspace = true
+#workspace = true
```
Comment (robert3005) [inline vortex-serde/Cargo.toml]: bring this back

**Comment by robert3005** [comment]
At this point the column reader is still not correct. Chunked and Flat layouts have the correct behaviour

File: vortex-expr/src/expr.rs [hunk 2]
```diff
@@ -8,24 +8,19 @@ use vortex::compute::{compare, Operator as ArrayOperator};
...
-
 pub trait VortexExpr: Debug + Send + Sync + PartialEq<dyn Any> {
     fn as_any(&self) -> &dyn Any;
 
     fn evaluate(&self, batch: &Array) -> VortexResult<Array>;
 
     fn references(&self) -> HashSet<Field>;
 
-    fn estimate_cost(&self, schema: &Schema) -> usize;
+    fn project(&self, projection: &[Field]) -> Option<Arc<dyn VortexExpr>>;
```
Comment (gatesn) [inline vortex-expr/src/expr.rs]: I don't understand what this function is supposed to do, or why it is on this trait?

Comment (robert3005) [reply vortex-expr/src/expr.rs]: The implementation is in #979. It's meant to project the filter into the requested column space or return empty if expression cannot be projected. This is very obscure way of naming but sort of made sense in my head. Not sure what a better name is

File: vortex-serde/src/layouts/read/mod.rs [hunk 3]
```diff
@@ -58,7 +74,15 @@ pub trait LayoutReader: Debug + Send {
     /// and then call back into this function.
     ///
     /// The layout is finished reading when it returns None
-    fn read_next(&mut self) -> VortexResult<Option<ReadResult>>;
+    fn read_next(&mut self, selection: RowSelector) -> VortexResult<Option<ReadResult>>;
+
+    /// Produce sets of row ranges to read from underlying layouts.
+    ///
+    /// Range terminating at end of the layout indicates that layout is done producing ranges.
+    fn read_range(&mut self) -> VortexResult<Option<RangeResult>>;
```
Comment (gatesn) [inline vortex-serde/src/layouts/read/mod.rs:83]: `next_range`

File: vortex-serde/src/layouts/read/mod.rs [hunk 4]
```diff
@@ -58,7 +74,15 @@ pub trait LayoutReader: Debug + Send {
...
+
+    /// Skip over next n_rows as we know they will not be part of result set
```
Comment (gatesn) [inline vortex-serde/src/layouts/read/mod.rs:85]: I think this doc string is wrong? We're advancing up to row N, not skipping over N rows

File: vortex-serde/src/layouts/read/mod.rs [hunk 5]
```diff
@@ -58,7 +74,15 @@ pub trait LayoutReader: Debug + Send {
...
+    fn advance(&mut self, up_to_row: usize) -> VortexResult<Messages>;
 
     // TODO(robert): Support stats pruning via planning. Requires propagating all the metadata
```
Comment (gatesn) [inline vortex-serde/src/layouts/read/mod.rs:88]: I think this is stale? Can remove

**Comment by robert3005** [comment]
Comments offline:
* there's just one method that takes expression
* add method from #979 for planning - call it sketch
* something about shortcircuiting siblings

**Comment by robert3005** [comment]
batching should be handled at range generation time. We shouldn't try to split read ranges into batches further

**Comment by robert3005** [comment]
revived as #1124

**Pull request closed without merging**